LibOS/shim/src/sys/shim_mmap.c

```diff
@@ -42,17 +42,25 @@ void * shim_do_mmap (void * addr, size_t length, int prot, int flags, int fd,
                      off_t offset)
 {
     struct shim_handle * hdl = NULL;
-    long ret = -ENOMEM;
-    bool reserved = false;
+    long ret = 0;
 
-    if (addr + length < addr) {
+    /*
+     * According to the manpage, both addr and offset have to be page-aligned,
+     * but not the length. mmap() will automatically round up the length.
+     */
+    if (addr && !ALIGNED(addr))
         return (void *) -EINVAL;
-    }
 
-    assert(!(flags & (VMA_UNMAPPED|VMA_TAINTED)));
+    if (fd >= 0 && !ALIGNED(offset))
+        return (void *) -EINVAL;
+
+    if (!ALIGNED(length))
+        length = ALIGN_UP(length);
 
-    if (flags & MAP_32BIT)
-        return (void *) -ENOSYS;
+    if (addr + length < addr)
```

Comment:
Overflow on pointers is UB, it is very likely that this `if` in the current form is optimized out from the machine code when using `-O2`.